src/mono/wasm/test-main.js

```diff
@@ -66,26 +66,30 @@ function proxyJson(func) {
         console[m] = proxyMethod(`console.${m}`, func, true);
 }
 
+var initConsole = Promise.resolve();
 if (is_browser) {
     const consoleUrl = `${window.location.origin}/console`.replace('http://', 'ws://');
 
     let consoleWebSocket = new WebSocket(consoleUrl);
     // redirect output so that when emscripten starts it's already redirected
-    proxyJson(function (msg) {
-        if (consoleWebSocket.readyState === WebSocket.OPEN) {
-            consoleWebSocket.send(msg);
-        }
-        else {
-            originalConsole.log(msg);
-        }
+    initConsole = new Promise(resolve => {
+        consoleWebSocket.onopen = function () {
+            proxyJson((msg) => {
+                consoleWebSocket.send(msg);
+            });
+            originalConsole.log("browser: Console websocket connected.");
+            resolve();
+        };
     });
 
-    consoleWebSocket.onopen = function (event) {
-        originalConsole.log("browser: Console websocket connected.");
-    };
     consoleWebSocket.onerror = function (event) {
         originalConsole.error(`websocket error: ${event}`);
     };
+
+    consoleWebSocket.onclose = function (event) {
```

Comment:
Somebody has to call close first in order for this to happen, no ?

Besides, if we are sending `WASM EXIT` on the same webSocket it will always  arrive after the full XML. 
WS is ordered protocol on top of TCP.

Reply:
the chrome bug I was looking at would cause the socket to close itself prematurely